File: libvbrief-ts/src/validation.ts
import {
  HIERARCHICAL_ID_PATTERN,
  ISSUE_DUPLICATE_ITEM_ID,
  ISSUE_INVALID_DOCUMENT_TYPE,
  ISSUE_INVALID_ID_FORMAT,
  ISSUE_INVALID_ITEM_STATUS,
  ISSUE_INVALID_ITEM_TYPE,
  ISSUE_INVALID_PLAN_FIELD_TYPE,
  ISSUE_INVALID_PLAN_STATUS,
  ISSUE_INVALID_PLANREF,
  ISSUE_INVALID_ROOT_FIELD_TYPE,
  ISSUE_INVALID_SUBITEMS_TYPE,
  ISSUE_INVALID_VERSION,
  ISSUE_MISSING_ITEM_FIELD,
  ISSUE_MISSING_PLAN_FIELD,
  ISSUE_MISSING_ROOT_FIELD,
  PLAN_REF_PATTERN,
  VALID_STATUSES,
} from "./compat.js";
import { validatePlanDAG } from "./dag.js";
import { ValidationReport } from "./issues.js";
import { isRecordLike } from "./utils.js";

export interface ValidateOptions {
  dag?: boolean;
}

/**
 * Validate a plain document or model object.
 */
export function validateDocument(document: unknown, options: ValidateOptions = {}): ValidationReport {
  const report = new ValidationReport();
  const data = toDict(document);

  if (!isRecordLike(data)) {
    report.addError(ISSUE_INVALID_DOCUMENT_TYPE, "$", "Document must be an object/dictionary");
    return report;
  }

  validateRoot(data, report);

  if (options.dag) {
    const plan = data.plan;
    if (isRecordLike(plan)) {
      validatePlanDAG(plan, report);
    }
  }

  return report;
}

function validateRoot(data: Record<string, unknown>, report: ValidationReport): void {
  let vbriefInfo: Record<string, unknown> | undefined;
  if (!Object.hasOwn(data, "vBRIEFInfo")) {
    report.addError(ISSUE_MISSING_ROOT_FIELD, "vBRIEFInfo", "Missing required root field: vBRIEFInfo");
  } else if (!isRecordLike(data.vBRIEFInfo)) {
    report.addError(ISSUE_INVALID_ROOT_FIELD_TYPE, "vBRIEFInfo", "vBRIEFInfo must be an object");
  } else {
    vbriefInfo = data.vBRIEFInfo;
  }

  if (vbriefInfo !== undefined) {
    const version = vbriefInfo.version;
    if (version !== "0.5") {
      report.addError(
        ISSUE_INVALID_VERSION,
        "vBRIEFInfo.version",
        `Expected version '0.5', got ${JSON.stringify(version)}`,
      );
    }
  }

  if (!Object.hasOwn(data, "plan")) {
    report.addError(ISSUE_MISSING_ROOT_FIELD, "plan", "Missing required root field: plan");
    return;
  }

  if (!isRecordLike(data.plan)) {
    report.addError(ISSUE_INVALID_ROOT_FIELD_TYPE, "plan", "plan must be an object");
    return;
  }

  validatePlan(data.plan, report);
}

function validatePlan(plan: Record<string, unknown>, report: ValidationReport): void {
  for (const fieldName of ["title", "status", "items"]) {
    if (!Object.hasOwn(plan, fieldName)) {
      report.addError(
        ISSUE_MISSING_PLAN_FIELD,
        `plan.${fieldName}`,
        `Missing required plan field: ${fieldName}`,
      );
    }
  }

  const status = plan.status;
  if (status !== undefined && (typeof status !== "string" || !VALID_STATUSES.has(status))) {
    report.addError(
      ISSUE_INVALID_PLAN_STATUS,
      "plan.status",
      `Invalid plan status ${JSON.stringify(status)}; expected one of ${JSON.stringify([...VALID_STATUSES].sort())}`,
    );
  }

  const planId = plan.id;
  if (planId !== undefined && (typeof planId !== "string" || !HIERARCHICAL_ID_PATTERN.test(planId))) {
    report.addError(ISSUE_INVALID_ID_FORMAT, "plan.id", "plan.id must match hierarchical ID pattern");
  }

  const items = plan.items;
  if (items === undefined) {
    return;
  }

  if (!Array.isArray(items)) {
    report.addError(ISSUE_INVALID_PLAN_FIELD_TYPE, "plan.items", "plan.items must be an array");
    return;
  }

  validateItems(items, report, "plan.items", new Set<string>());
}

function validateItems(
  items: unknown[],
  report: ValidationReport,
  path: string,
  seenIds: Set<string>,
): void {
  for (const [index, item] of items.entries()) {
    const itemPath = `${path}[${index}]`;

    if (!isRecordLike(item)) {
      report.addError(ISSUE_INVALID_ITEM_TYPE, itemPath, "Plan item must be an object");
      continue;
    }

    if (!Object.hasOwn(item, "title")) {
      report.addError(ISSUE_MISSING_ITEM_FIELD, `${itemPath}.title`, "Missing required item field: title");
    }

    if (!Object.hasOwn(item, "status")) {
      report.addError(ISSUE_MISSING_ITEM_FIELD, `${itemPath}.status`, "Missing required item field: status");
    }

    const status = item.status;
    if (status !== undefined && (typeof status !== "string" || !VALID_STATUSES.has(status))) {
      report.addError(
        ISSUE_INVALID_ITEM_STATUS,
        `${itemPath}.status`,
        `Invalid item status ${JSON.stringify(status)}; expected one of ${JSON.stringify([...VALID_STATUSES].sort())}`,
      );
    }

    const itemId = item.id;
    if (itemId !== undefined && (typeof itemId !== "string" || !HIERARCHICAL_ID_PATTERN.test(itemId))) {
      report.addError(ISSUE_INVALID_ID_FORMAT, `${itemPath}.id`, "item id must match hierarchical ID pattern");
    } else if (typeof itemId === "string") {
      if (seenIds.has(itemId)) {
        report.addError(ISSUE_DUPLICATE_ITEM_ID, `${itemPath}.id`, `Duplicate item id ${JSON.stringify(itemId)}`);
      } else {
        seenIds.add(itemId);
      }
    }

    const planRef = item.planRef;
    if (planRef !== undefined && (typeof planRef !== "string" || !PLAN_REF_PATTERN.test(planRef))) {
      report.addError(
        ISSUE_INVALID_PLANREF,
        `${itemPath}.planRef`,
        "planRef must match #..., file://..., or https://...",
      );
    }

    const subItems = item.subItems;
    if (subItems === undefined) {
      continue;
    }
    if (!Array.isArray(subItems)) {
      report.addError(ISSUE_INVALID_SUBITEMS_TYPE, `${itemPath}.subItems`, "subItems must be an array");
      continue;
    }

    validateItems(subItems, report, `${itemPath}.subItems`, seenIds);
  }
}

function toDict(document: unknown): unknown {

  if (typeof document === "object" && document !== null) {
    const candidate = document as { toDict?: (options?: { preserveOrder?: boolean }) => unknown };
    if (typeof candidate.toDict === "function") {
      return candidate.toDict({ preserveOrder: false });
    }
  }

  if (isRecordLike(document)) {
    return document;
  }

  return document;
}

File: libvbrief/validation.py
"""Core conformance validation for vBRIEF v0.5 JSON documents."""

from __future__ import annotations

from typing import Any, Mapping

from libvbrief.compat import (
    HIERARCHICAL_ID_PATTERN,
    ISSUE_DUPLICATE_ITEM_ID,
    ISSUE_INVALID_DOCUMENT_TYPE,
    ISSUE_INVALID_ID_FORMAT,
    ISSUE_INVALID_ITEM_STATUS,
    ISSUE_INVALID_ITEM_TYPE,
    ISSUE_INVALID_PLANREF,
    ISSUE_INVALID_PLAN_FIELD_TYPE,
    ISSUE_INVALID_PLAN_STATUS,
    ISSUE_INVALID_ROOT_FIELD_TYPE,
    ISSUE_INVALID_SUBITEMS_TYPE,
    ISSUE_INVALID_VERSION,
    ISSUE_MISSING_ITEM_FIELD,
    ISSUE_MISSING_PLAN_FIELD,
    ISSUE_MISSING_ROOT_FIELD,
    PLAN_REF_PATTERN,
    VALID_STATUSES,
)
from libvbrief.issues import ValidationReport


def validate_document(document: Any, *, dag: bool = False) -> ValidationReport:
    """Validate dict or model document and return structured issues.

    Pass ``dag=True`` to also check that plan.edges form a DAG (no dangling
    references and no cycles).  DAG checks are skipped when ``dag=False``
    (the default) to preserve backward compatibility.
    """
    report = ValidationReport()
    data = _to_dict(document)

    if not isinstance(data, Mapping):
        report.add_error(
            ISSUE_INVALID_DOCUMENT_TYPE,
            "$",
            "Document must be an object/dictionary",
        )
        return report

    _validate_root(data, report)

    if dag:
        plan = data.get("plan")
        if isinstance(plan, Mapping):
            from libvbrief.dag import validate_plan_dag
            validate_plan_dag(plan, report)

    return report


def _validate_root(data: Mapping[str, Any], report: ValidationReport) -> None:
    if "vBRIEFInfo" not in data:
        report.add_error(ISSUE_MISSING_ROOT_FIELD, "vBRIEFInfo", "Missing required root field: vBRIEFInfo")
        vbrief_info: Mapping[str, Any] | None = None
    else:
        raw_info = data.get("vBRIEFInfo")
        if not isinstance(raw_info, Mapping):
            report.add_error(
                ISSUE_INVALID_ROOT_FIELD_TYPE,
                "vBRIEFInfo",
                "vBRIEFInfo must be an object",
            )
            vbrief_info = None
        else:
            vbrief_info = raw_info

    if vbrief_info is not None:
        version = vbrief_info.get("version")
        if version != "0.5":
            report.add_error(
                ISSUE_INVALID_VERSION,
                "vBRIEFInfo.version",
                f"Expected version '0.5', got {version!r}",
            )

    if "plan" not in data:
        report.add_error(ISSUE_MISSING_ROOT_FIELD, "plan", "Missing required root field: plan")
        return

    plan = data.get("plan")
    if not isinstance(plan, Mapping):
        report.add_error(ISSUE_INVALID_ROOT_FIELD_TYPE, "plan", "plan must be an object")
        return

    _validate_plan(plan, report)


def _validate_plan(plan: Mapping[str, Any], report: ValidationReport) -> None:
    for field_name in ("title", "status", "items"):
        if field_name not in plan:
            report.add_error(
                ISSUE_MISSING_PLAN_FIELD,
                f"plan.{field_name}",
                f"Missing required plan field: {field_name}",
            )

    status = plan.get("status")
    if status is not None and status not in VALID_STATUSES:
        report.add_error(
            ISSUE_INVALID_PLAN_STATUS,
            "plan.status",
            f"Invalid plan status {status!r}; expected one of {sorted(VALID_STATUSES)}",
        )

    plan_id = plan.get("id")
    if plan_id is not None and (not isinstance(plan_id, str) or not HIERARCHICAL_ID_PATTERN.match(plan_id)):
        report.add_error(
            ISSUE_INVALID_ID_FORMAT,
            "plan.id",
            "plan.id must match hierarchical ID pattern",
        )

    items = plan.get("items")
    if items is None:
        return

    if not isinstance(items, list):
        report.add_error(
            ISSUE_INVALID_PLAN_FIELD_TYPE,
            "plan.items",
            "plan.items must be an array",
        )
        return

    _validate_items(items, report, "plan.items", seen_ids=set())

def _validate_items(
    items: list[Any],
    report: ValidationReport,
    path: str,
    *,
    seen_ids: set[str],
) -> None:
    for index, item in enumerate(items):
        item_path = f"{path}[{index}]"

        if not isinstance(item, Mapping):
            report.add_error(
                ISSUE_INVALID_ITEM_TYPE,
                item_path,
                "Plan item must be an object",
            )
            continue

        if "title" not in item:
            report.add_error(
                ISSUE_MISSING_ITEM_FIELD,
                f"{item_path}.title",
                "Missing required item field: title",
            )

        if "status" not in item:
            report.add_error(
                ISSUE_MISSING_ITEM_FIELD,
                f"{item_path}.status",
                "Missing required item field: status",
            )

        status = item.get("status")
        if status is not None and status not in VALID_STATUSES:
            report.add_error(
                ISSUE_INVALID_ITEM_STATUS,
                f"{item_path}.status",
                f"Invalid item status {status!r}; expected one of {sorted(VALID_STATUSES)}",
            )

        item_id = item.get("id")
        if item_id is not None and (not isinstance(item_id, str) or not HIERARCHICAL_ID_PATTERN.match(item_id)):
            report.add_error(
                ISSUE_INVALID_ID_FORMAT,
                f"{item_path}.id",
                "item id must match hierarchical ID pattern",
            )
        elif isinstance(item_id, str):
            if item_id in seen_ids:
                report.add_error(
                    ISSUE_DUPLICATE_ITEM_ID,
                    f"{item_path}.id",
                    f"Duplicate item id {item_id!r}",
                )
            else:
                seen_ids.add(item_id)

        plan_ref = item.get("planRef")
        if plan_ref is not None and (not isinstance(plan_ref, str) or not PLAN_REF_PATTERN.match(plan_ref)):
            report.add_error(
                ISSUE_INVALID_PLANREF,
                f"{item_path}.planRef",
                "planRef must match #..., file://..., or https://...",
            )

        sub_items = item.get("subItems")
        if sub_items is None:
            continue
        if not isinstance(sub_items, list):
            report.add_error(
                ISSUE_INVALID_SUBITEMS_TYPE,
                f"{item_path}.subItems",
                "subItems must be an array",
            )
            continue

        _validate_items(sub_items, report, f"{item_path}.subItems", seen_ids=seen_ids)


def _to_dict(document: Any) -> Any:
    if isinstance(document, Mapping):
        return document

    to_dict = getattr(document, "to_dict", None)
    if callable(to_dict):
        return to_dict(preserve_order=False)

    return document
